File: src/main/org/cuair/ground/models/Image.java
package org.cuair.ground.models;

import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.HashMap;
import java.util.Map;
import javax.persistence.Entity;
import javax.persistence.OneToOne;
import org.cuair.ground.models.geotag.FOV;
import org.cuair.ground.models.geotag.GpsLocation;
import org.cuair.ground.models.geotag.Telemetry;
import org.cuair.ground.util.Geotagging;

@Entity
public class Image extends TimestampModel {

  /** The URL where clients can retrieve the image file  */
  private String imageUrl;

  /** The local URL where image file lives on the ground server */
  private String localImageUrl;

  /** Closest Telemetry for when this Image was taken  */
  private Telemetry telemetry;

  /** The type of this image. Either FIXED, TRACKING, or OFFAXIS */
  private ImgMode imgMode;

  /** True if has at least one associated MDLC assignment, otherwise false. */
  private boolean hasMdlcAssignment = false;

  /** True if has at least one associated ADLC assignment, otherwise false. */
  private boolean hasAdlcAssignment = false;

  /** The field of view of this image. */
  private FOV fov;

  /** The possible image modes: fixed, tracking, and off-axis */
  public enum ImgMode {
    FIXED("fixed"),
    TRACKING("tracking"),
    OFFAXIS("off-axis");

    // either add @JsonValue here (if you don't need getter)
    @JsonValue String value;

    ImgMode(String value) { this.value = value; }
  }

  public Image (String imageUrl, Telemetry telemetry, FOV fov, ImgMode imgMode) {
    this.imageUrl = imageUrl;
    this.telemetry = telemetry;
    this.fov = fov;
    this.imgMode = imgMode;
  }

                                    /**
   * Internal method for finding geotags corresponding to four corners of image
   */
  public Map<String, Object> getLocations() {
    GpsLocation imageGPS = telemetry.getGps();
    double centerLat = imageGPS.getLatitude();
    double centerLong = imageGPS.getLongitude();

    double planeYaw = telemetry.getPlaneYaw();
    double altitude = telemetry.getAltitude();

    GpsLocation topLeft = Geotagging
        .getPixelCoordinates(centerLat, centerLong, altitude, fov, 0.0, 0.0, planeYaw);
    GpsLocation topRight = Geotagging
        .getPixelCoordinates(centerLat, centerLong, altitude, fov, Geotagging.IMAGE_WIDTH, 0.0, planeYaw);
    GpsLocation bottomLeft = Geotagging
        .getPixelCoordinates(centerLat, centerLong, altitude, fov, 0.0, Geotagging.IMAGE_HEIGHT, planeYaw);
    GpsLocation bottomRight = Geotagging
        .getPixelCoordinates(centerLat, centerLong, altitude, fov, Geotagging.IMAGE_WIDTH, Geotagging.IMAGE_HEIGHT, planeYaw);

    Map<String, Object> locs = new HashMap<>();
    locs.put("topLeft", topLeft);
    locs.put("topRight", topRight);
    locs.put("bottomLeft", bottomLeft);
    locs.put("bottomRight", bottomRight);

    return locs;
  }

  public String getImageUrl() {
    return imageUrl;
  }

  public String getLocalImageUrl() {
    return localImageUrl;
  }

  public Telemetry getTelemetry() {
    return telemetry;
  }

  public ImgMode getImgMode() {
    return imgMode;
  }

  public FOV getFov() {
    return fov;
  }

  public void setLocalImageUrl(String url) {
    localImageUrl = url;
  }

  public void setHasMdlcAssignment(boolean has) {
    hasMdlcAssignment = has;
  }

  public void setHasAdlcAssignment(boolean has) {
    hasAdlcAssignment = has;
  }

  public void setImageUrl(String imageUrl) {
    this.imageUrl = imageUrl;
  }

  public void setTelemetry(Telemetry telemetry) {
    this.telemetry = telemetry;
  }

  public void setImgMode(ImgMode imgMode) {
    this.imgMode = imgMode;
  }

  public void setFov(FOV fov) {
    this.fov = fov;
  }


}

File: src/main/org/cuair/ground/models/Image.kt
package org.cuair.ground.models

import com.fasterxml.jackson.annotation.JsonIgnore
import org.cuair.ground.models.geotag.Telemetry
import org.cuair.ground.util.Geotagging
import java.util.Objects
import javax.persistence.Entity
import javax.persistence.OneToOne
import javax.persistence.CascadeType

import io.ebean.annotation.EnumValue
import com.fasterxml.jackson.annotation.JsonProperty

/** Represents an image and its corresponding metadata as sent down from the plane */
@Entity
class Image(
        /** The URL where clients can retrieve the image file  */
        var imageUrl: String?,
        /** Closest Telemetry for when this Image was taken  */
        @OneToOne(cascade = [CascadeType.ALL])
        var telemetry: Telemetry,
        /** The type of this image. Either FIXED, TRACKING, or OFFAXIS */
        var imgMode: ImgMode,
        /** True if has at least one associated MDLC assignment, otherwise false. */
        @JsonIgnore var hasMdlcAssignment: Boolean = false,
        /** True if has at least one associated ADLC assignment, otherwise false. */
        @JsonIgnore var hasAdlcAssignment: Boolean = false,
        /** The horizontal field of fiew of this image in degrees. */
        var fov: Double
) : TimestampModel() {

    /** Secondary constructor that sets imgMode to FIXED by default */
    constructor(
            imageUrl: String,
            telemetry: Telemetry,
            fov: Double
    ) : this(imageUrl, telemetry, ImgMode.FIXED, false, false, fov)

    /** The filesystem path this image lives on relative to the server directory */
    @Transient
    var localImageUrl: String? = null

    /** The possible image modes: fixed, tracking, and off-axis */
    enum class ImgMode(val mode: String) {
        @JsonProperty("fixed")
        @EnumValue("0")
        FIXED("fixed"),
        @JsonProperty("tracking")
        @EnumValue("1")
        TRACKING("tracking"),
        @JsonProperty("off-axis")
        @EnumValue("2")
        OFFAXIS("off-axis")
    }

    /**
     * Internal method for finding geotags for a given image id
     *
     * @return ObjectNode
     */
    @JsonIgnore
    fun getLocations(): MutableMap<String, Any?>? {
        val imageTelemetry = this.telemetry ?: return null
        val imageGPS = imageTelemetry.getGps() ?: return null
        val centerLatitude = imageGPS.getLatitude()
        val centerLongitude = imageGPS.getLongitude()
        if (imageTelemetry.getPlaneYaw() == null) return null
        var planeYaw = imageTelemetry.getPlaneYaw() * Math.PI / 180
        val altitude = imageTelemetry.getAltitude() ?: return null

        val topLeft = Geotagging.getPixelCoordinates(centerLatitude, centerLongitude, altitude, 0.0, 0.0, planeYaw)
        val topRight = Geotagging.getPixelCoordinates(
                centerLatitude, centerLongitude, altitude, Geotagging.IMAGE_WIDTH, 0.0, planeYaw)
        val bottomLeft = Geotagging.getPixelCoordinates(
                centerLatitude, centerLongitude, altitude, 0.0, Geotagging.IMAGE_HEIGHT, planeYaw)
        val bottomRight = Geotagging.getPixelCoordinates(
                centerLatitude,
                centerLongitude,
                altitude,
                Geotagging.IMAGE_WIDTH,
                Geotagging.IMAGE_HEIGHT,
                planeYaw)

        val locs = mutableMapOf<String, Any?>()
        locs.put("topLeft", topLeft)
        locs.put("topRight", topRight)
        locs.put("bottomLeft", bottomLeft)
        locs.put("bottomRight", bottomRight)
        locs.put("orientation", planeYaw)
        locs.put("url", this.imageUrl)

        return locs
    }

    /**
     * Determines if the given object is logically equal to this Image
     *
     * @param other The object to compare
     * @return true if the object equals this Image
     */
    override fun equals(other: Any?): Boolean {
        if (other !is Image) return false

        if (this.imageUrl != other.imageUrl) return false
        if (this.imgMode != other.imgMode) return false
        if (!Objects.equals(this.telemetry, other.telemetry)) return false
        if (this.fov != other.fov) return false
        return true
    }
}
